crates/sui-faucet/src/faucet/simple_faucet.rs

```diff
@@ -103,57 +110,71 @@ impl SimpleFaucet {
                 error!(?uuid, "Timeout when getting consumer lock");
                 Err(anyhow::anyhow!("Too many concurrent requests, try later!"))
             }
-        }?;
-        info!(?uuid, "Got consumer lock, pulling coins.");
-        loop {
-            match tokio::time::timeout(
-                Duration::from_secs(WAIT_FOR_GAS_TIMEOUT_SEC),
-                consumer.recv(),
-            )
-            .await
-            {
-                Ok(Some(coin)) => {
-                    info!(?uuid, "Pulling coin from pool {:?}", coin);
-                    self.metrics.total_available_coins.dec();
-                    let gas_coin = self.get_gas_coin(coin).await?;
-                    if let Some(gas_coin) = gas_coin {
-                        if gas_coin.value() >= total_amount + PAY_SUI_GAS {
-                            info!(
-                                ?uuid,
-                                "Planning to use coin from pool {:?}, current balance: {}",
-                                coin,
-                                gas_coin.value()
-                            );
-                            return Ok(coin);
+        };
+        if let Ok(mut consumer) = consumer_res {
+            info!(?uuid, "Got consumer lock, pulling coins.");
+            loop {
+                match tokio::time::timeout(
+                    Duration::from_secs(WAIT_FOR_GAS_TIMEOUT_SEC),
+                    consumer.recv(),
+                )
+                .await
+                {
+                    Ok(Some(coin)) => {
+                        info!(?uuid, "Pulling coin from pool {:?}", coin);
+                        self.metrics.total_available_coins.dec();
+                        let gas_coin = self.get_gas_coin(coin).await;
+                        if let Ok(gas_coin_opt) = gas_coin {
```

Comment:
A `match` with guards might be easier to follow here:

```
match self.get_gas_coin(coin).await {
  Ok(Some(gas_coin)) if gas_coin.value() >= total_amount + PAY_SUI_GAS => {
    // ...
  }

  Ok(Some(_)) => {
    // ... remove it from the pool, insufficient balance
  }

  Ok(None) => {
    // ... invalid gas
  }

  Err(_)=> {
    // ... unknown gas coin.
  }
}
```

Reply:
This has highlighted by the way that this code currently doesn't do anything with the error value from the request, but that's probably worth doing, either printing it or re-packaging it in the return value so that it can be propagated further up the stack after the recycling has happened.